server/sonar-server/src/main/java/org/sonar/server/qualityprofile/DefinedQProfile.java

```diff
@@ -0,0 +1,127 @@
+import org.sonar.api.profiles.ProfileDefinition;
+
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.apache.commons.codec.binary.Hex.encodeHexString;
+import static org.apache.commons.lang.StringUtils.lowerCase;
+import static org.sonar.db.loadedtemplate.LoadedTemplateDto.QUALITY_PROFILE_TYPE;
+
+/**
+ * Represent a Quality Profile as computed from {@link ProfileDefinition} provided by installed plugins.
+ */
+public final class DefinedQProfile {
+  private final QProfileName qProfileName;
+  private final boolean isDefault;
+  private final String loadedTemplateType;
+  private final List<org.sonar.api.rules.ActiveRule> activeRules;
+
+  private DefinedQProfile(Builder builder, MessageDigest messageDigest) {
+    this.qProfileName = new QProfileName(builder.getLanguage(), builder.getName());
+    this.isDefault = builder.declaredDefault || builder.computedDefault;
+    this.loadedTemplateType = computeLoadedTemplateType(this.qProfileName, messageDigest);
+    this.activeRules = ImmutableList.copyOf(builder.activeRules);
+  }
+
+  private static String computeLoadedTemplateType(QProfileName qProfileName, MessageDigest messageDigest) {
+    String qpIdentifier = lowerCase(qProfileName.getLanguage()) + ":" + qProfileName.getName();
+    return format("%s.%s", QUALITY_PROFILE_TYPE, encodeHexString(messageDigest.digest(qpIdentifier.getBytes(UTF_8))));
+  }
+
+  public String getName() {
+    return qProfileName.getName();
+  }
+
+  public String getLanguage() {
+    return qProfileName.getLanguage();
+  }
+
+  public QProfileName getQProfileName() {
+    return qProfileName;
+  }
+
+  public boolean isDefault() {
+    return isDefault;
+  }
+
+  public String getLoadedTemplateType() {
+    return loadedTemplateType;
+  }
+
+  public List<org.sonar.api.rules.ActiveRule> getActiveRules() {
+    return activeRules;
+  }
+
+  static final class Builder {
+    private String language;
+    private String name;
+    private boolean declaredDefault;
+    private boolean computedDefault;
+    private List<org.sonar.api.rules.ActiveRule> activeRules = new ArrayList<>();
+
+    public String getLanguage() {
+      return language;
+    }
+
+    public Builder setLanguage(String language) {
+      this.language = language;
+      return this;
+    }
+
+    Builder setName(String name) {
+      this.name = name;
+      return this;
+    }
+
+    String getName() {
+      return name;
+    }
+
+    Builder setDeclaredDefault(boolean declaredDefault) {
+      this.declaredDefault = declaredDefault;
+      return this;
+    }
+
+    boolean isDeclaredDefault() {
+      return declaredDefault;
+    }
+
+    Builder setComputedDefault(boolean flag) {
+      computedDefault = flag;
+      return this;
+    }
+
+    Builder addRules(List<org.sonar.api.rules.ActiveRule> rules) {
+      this.activeRules.addAll(rules);
+      return this;
+    }
+
+    DefinedQProfile build(MessageDigest messageDigest) {
```

Comment:
What's the point for making MessageDigest configurable? All instances of `DefinedQProfile` use md5.